data_pipeline/twitter_ads_api/twitter_ads_api_pipeline.py

```diff
@@ -85,36 +84,87 @@ def get_param_dict_from_api_query_parameters(
+) -> datetime:
+    yesterday_date = get_yesterdays_date()
+    end_date_by_period = initial_start_date_value + timedelta(
+        days=api_query_parameters_config.parameter_values.ending_period_per_day
+    )
+    if end_date_by_period > yesterday_date:
+        final_end_date_value = yesterday_date
+    else:
+        final_end_date_value = end_date_by_period
+    return final_end_date_value
+
+
+def get_end_date_value_of_period(
+    start_date_value: datetime,
+    final_end_date_value: datetime,
+    days_in_period: int
+) -> datetime:
+    period_end_date_value = start_date_value + timedelta(days=days_in_period)
+    if period_end_date_value > final_end_date_value:
+        end_date_value = final_end_date_value
+    else:
+        end_date_value = period_end_date_value
+    return end_date_value
+
+
 def iter_bq_compatible_json_response_from_resource_with_provenance(
     source_config: TwitterAdsApiSourceConfig
 ) -> Any:
     if source_config.api_query_parameters:
         api_query_parameters_config = source_config.api_query_parameters
-        placement_value_list = api_query_parameters_config.parameter_values.placement_value
-        if api_query_parameters_config.parameter_names_for.placement:
-            assert placement_value_list
-            assert api_query_parameters_config.use_start_date_from_bigquery
-            dict_value_list_from_bq = list(iter_dict_from_bq_query_for_bigquery_source_config(
-                api_query_parameters_config.parameter_values.from_bigquery
-            ))
-            LOGGER.debug("dict_value_list_from_bq: %r", dict_value_list_from_bq)
-            for dict_value_from_bq in dict_value_list_from_bq:
+        dict_value_list_from_bq = list(iter_dict_from_bq_query_for_bigquery_source_config(
+            api_query_parameters_config.parameter_values.from_bigquery
+        ))
+        LOGGER.debug("dict_value_list_from_bq: %r", dict_value_list_from_bq)
+        for dict_value_from_bq in dict_value_list_from_bq:
+            entity_id_from_bq = dict_value_from_bq['entity_id']
+            start_date_str_from_bq = dict_value_from_bq['start_date']
+            initial_start_date_value = (
+                datetime.strptime(start_date_str_from_bq, '%Y-%m-%d').date()
+            )
+            if initial_start_date_value <= datetime.strptime('2015-01-01', '%Y-%m-%d').date():
+                LOGGER.info(
+                    "initial_start_date %s for entity_id %s is out of range for the API",
+                    start_date_str_from_bq,
+                    entity_id_from_bq
+                )
+                continue
+            final_end_date_value = get_final_end_date(
+                api_query_parameters_config=api_query_parameters_config,
+                initial_start_date_value=initial_start_date_value
+            )
+
+            placement_value_list = api_query_parameters_config.parameter_values.placement_value
+            if not api_query_parameters_config.parameter_names_for.placement:
```

Comment:
My reading of this is that the assumption is that if we do not have a placement value, then we do not need to break up the period. It might be useful to add an assert that `ending_period_per_day` (or the final name) was not specified.

Reply:
Actually thinking about it again, it is not `ending_period_per_day`. It is the `7 days` that isn't configured.

Reply:
`batch_size_in_days`